### PR TITLE
add capsule, some typos, separate checks u and v

### PR DESCRIPTION
**Pull request by worbit**
### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.

As discussed, add the `Capsule` class, will add unimplemented methods (eg `to_vertices_and_faces`)

### CODE REVIEW

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    @property
+    def line(self):
+        return self._line
+    
+    @line.setter
+    def line(self, line):
+        # add check if segment is
+        # ((x,y,z),(x,y,z)) or
+        # [[x,y,z],[x,y,z]] or
+        # [point,point] or ...
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/shapes/capsule.py]: what is all this for?

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        capsule.line.transform(matrix)
+        return capsule
+
+
+# ==============================================================================
+# Main
+# ==============================================================================
+
+if __name__ == "__main__":
+    pass
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/shapes/capsule.py]: end with newline...

**Review by tomvanmele — COMMENTED** [review]
i guess the only thing left to do is the caps?

File: src/compas/geometry/primitives/shapes/capsule.py (hunk 1 above)
Comment (worbit) [reply src/compas/geometry/primitives/shapes/capsule.py]: thanks for reminding me, this is a leftover from before i used a `Line` object for the `Capsule` axis

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            The transformation matrix.
+        """
+        self.line.transform(matrix)
+    
+    def transformed(self, matrix):
+        """Return a transformed copy of this ``Capsule`` using a given transformation matrix.
+
+        Parameters
+        ----------
+        matrix : list of list
```
Comment (romanarust) [inline src/compas/geometry/primitives/shapes/capsule.py]: Shouldn't this be rather a :class:`Transformation` than the pure matrix?

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            The transformation matrix.
+
+        Returns
+        -------
+        Capsule
```
Comment (romanarust) [inline src/compas/geometry/primitives/shapes/capsule.py]:
```suggestion
:class:`Capsule`
```

File: src/compas/geometry/primitives/line.py [hunk 5]
```diff
@@ -260,7 +260,7 @@ def copy(self):
     # ==========================================================================
 
     def transform(self, matrix):
```
Comment (romanarust) [inline src/compas/geometry/primitives/line.py]: shouldn't this be rather a :class:`Transformation`?

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            The copy.
+        """
+        cls = type(self)
+        return cls(self.line.copy(), self.height)
+
+    # ==========================================================================
+    # transformations
+    # ==========================================================================
+
+    def transform(self, matrix):
```
Comment (romanarust) [inline src/compas/geometry/primitives/shapes/capsule.py]:
```suggestion
    def transform(self, transformation):
```

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    Attributes
+    ----------
+    segment: :class:`compas.geometry.Line`
+        The axis line of the capsule.
+    radius: float
+        The radius of the capsule.
+
+    Examples
+    --------
+    >>> TODO
```
Comment (romanarust) [inline src/compas/geometry/primitives/shapes/capsule.py]:
```suggestion
    >>> line = Line((1, 2, 3), (5, 4, 3))
    >>> capsule = Capsule(line, 1.2)
```

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        """Returns the data dictionary that represents the capsule.
+
+        Returns
+        -------
+        dict
+            The capsule data.
+
+        """
+        return self.data
+
```
Comment (romanarust) [inline src/compas/geometry/primitives/shapes/capsule.py:115]: what about a `__repr__` that returns a string that can be used to construct a Capsule again, like:
```suggestion
    def __repr__(self):
        return "Capsule({0}, {1})".format(self.line, self.radius)
```

**Review by romanarust — CHANGES_REQUESTED** [review]
docstrings and examples missing here and there

File: src/compas/geometry/primitives/line.py (hunk 5 above)
Comment (tomvanmele) [reply src/compas/geometry/primitives/line.py]: yes indeed

File: src/compas/geometry/primitives/shapes/capsule.py (hunk 8 above)
Comment (tomvanmele) [reply src/compas/geometry/primitives/shapes/capsule.py:115]: does `repr` get applied recursively? i mean, will this apply `repr` to `self.line` as well?

File: src/compas/geometry/primitives/shapes/capsule.py (hunk 6 above)
Comment (tomvanmele) [reply src/compas/geometry/primitives/shapes/capsule.py]: agreed

File: src/compas/geometry/primitives/shapes/capsule.py (hunk 3 above)
Comment (tomvanmele) [reply src/compas/geometry/primitives/shapes/capsule.py]: yes indeed

File: src/compas/geometry/primitives/shapes/capsule.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+# Main
+# ==============================================================================
+
+if __name__ == "__main__":
+    from compas.datastructures import Mesh
+    line = Line((1, 2, 3), (5, 4, 3))
+    capsule = Capsule(line, 1.2)
+
+    mesh = Mesh.from_shape(capsule)
+    mesh.to_obj('/Users/bernham/Desktop/capsule.obj')
```
Comment (tomvanmele) [inline src/compas/geometry/primitives/shapes/capsule.py]: should this not be a call to `doctest`?

File: src/compas/geometry/primitives/shapes/capsule.py (hunk 3 above)
Comment (worbit) [reply src/compas/geometry/primitives/shapes/capsule.py]: Agree. I copied this from `primitives.Line` where it is also a matrix and this same docstring. I will go through all the `Shape` classes and update it according to your suggestion.

**Comment by tomvanmele** [comment]
@worbit is this thing intentionally still a draft?

**Comment by worbit** [comment]
no, i think it's good to get merged!

**Pull request merged**